### PR TITLE
2021.02

### PR DESCRIPTION
**Pull request by antgonza**

### CODE REVIEW

**Comment by antgonza** [comment]
This will fail until https://github.com/qiita-spots/qtp-diversity/pull/13 is merged; so review/merge that one first.

File: qp_qiime2/__init__.py [hunk 1]
```diff
@@ -67,7 +75,9 @@
             qiita_artifact = 'BIOM'
 
         if q2plugin.name not in Q2_ALLOWED_PLUGINS:
-            # This currently filters (which are processing commands):
+            # As of qiime2-2021.2 this filters out:
+            # diversity-lib
+            # empress
```
Comment (ElDeveloper) [inline qp_qiime2/__init__.py:80]: 🌳

**Review by ElDeveloper — APPROVED** [review]
I'm not exactly following what happened in `qp_qiime2/__init__.py` any hints there? Seems like most of the code is the same but things were shuffled around.

File: qp_qiime2/__init__.py [hunk 2]
```diff
@@ -38,21 +39,28 @@
...
-# 2. We are going to "force" add commands that are not added in the previous
-# loop based on user demand
 #
 # Note that Qiita users like to have descriptions of the paramters
 # (q2-description) vs. the parameter itself (q2-parameter) so to allow this
 # we are going to store each parameter twice: one in the
 # opt_params[q2-description]: value; and
 # req_params['qp-hide-param' + q2-description]: q2-parameter
 
+pm = PluginManager()
```
Comment (antgonza) [inline qp_qiime2/__init__.py:57]: This block, up to line 62, is the replacement of all the code below 278 to add `methods['classify_sklearn']`. I guess this highlights the spirit of the changes: realized that much of the code was duplicated to add that method and that we could simplify it if we first just create a list of the methods we will add (`methods_to_add`) and then add them.

File: qp_qiime2/__init__.py [hunk 3]
```diff
@@ -38,21 +39,28 @@
...
+methods_to_add = []
+for plugin_name, method_name in Q2_EXTRA_COMMANDS:
```
Comment (antgonza) [inline qp_qiime2/__init__.py:59]: Here we loop the methods that we are adding by hand.

File: qp_qiime2/__init__.py [hunk 4]
```diff
@@ -38,21 +39,28 @@
...
+    q2plugin = pm.plugins[plugin_name]
+    m = q2plugin.methods[method_name]
+    methods_to_add.append((q2plugin, m))
+
 for qiita_artifact, q2_artifacts in QIITA_Q2_SEMANTIC_TYPE.items():
```
Comment (antgonza) [inline qp_qiime2/__init__.py:64]: Here the methods we are gonna source from the install.

File: qp_qiime2/__init__.py [hunk 5]
```diff
@@ -76,234 +86,191 @@
...
-                continue
+            if (q2plugin.name, m.id) not in [('feature-table', 'group'),
+                                             ('feature-table', 'filter_seqs')]:
+                methods_to_add.append((q2plugin, m))
 
-            inputs = m.signature.inputs.copy()
-            outputs = m.signature.outputs.copy()
-            parameters = m.signature.parameters.copy()
-            add_method = True
+for q2plugin, m in methods_to_add:
```
Comment (antgonza) [inline qp_qiime2/__init__.py:93]: Once we have all the `methods_to_add `, we simply do what we used to do in the block above - however, the code and ident had to slightly change to  accommodate the new structure.

File: qp_qiime2/__init__.py [hunk 6]
```diff
@@ -76,234 +86,191 @@
...
+    inputs = m.signature.inputs.copy()
```
Comment (antgonza) [inline qp_qiime2/__init__.py:94]: For example, for the ident, these block is the same than in the old 83-86 block with less tabs.

File: qp_qiime2/__init__.py [hunk 7]
```diff
@@ -76,234 +86,191 @@
...
-                else:
-                    ename = element.description
-                    req_params[ename] = ('artifact', [etype])
-                # we need to add the actual name of the parameter so we
-                # can retrieve later
-                req_params['qp-hide-param' + ename] = ('string', pname)
-
-            for td in to_delete:
-                del inputs[td]
+    # storing this information in req_params so we can use internally
```
Comment (antgonza) [inline qp_qiime2/__init__.py:99]: Another example of ident, these block used to be 88-94

File: qp_qiime2/__init__.py [hunk 8]
```diff
@@ -76,234 +86,191 @@
...
+        add_method = True
+        etype = Q2_QIITA_SEMANTIC_TYPE[qt_name]
+        # these are special types as we can retrive internally
+        if etype in 'phylogeny':
+            ename = 'Phylogenetic tree'
+            req_params[ename] = (
+                'choice:["None", "Artifact tree, if exists"]', 'None')
+            req_params['qp-hide-param' + ename] = ('string', pname)
+            to_delete.append(pname)
+        elif etype == 'FeatureData':
```
Comment (antgonza) [inline qp_qiime2/__init__.py:122]: These is the only "real" change to support the change in https://github.com/qiita-spots/qtp-diversity/pull/13 and allow the `predicate == ['Sequence']` block.

File: qp_qiime2/__init__.py [hunk 9]
```diff
@@ -76,234 +86,191 @@
...
-                # if we are in the diversity plugin, the method starts with
-                # alpha/beta, and the parameter is 'metric', we might want
-                # to replace the technical names for user friendly ones
-                value_pair = (mid, pname)
-                if qname == 'diversity' and value_pair in RENAME_COMMANDS:
-                    # converting to list to the serialize doesn't complaint
-                    vals = list(RENAME_COMMANDS[value_pair])
-                    data_type = 'choice:%s' % dumps(vals)
-                    default = vals[0]
+    if add_method:
```
Comment (antgonza) [inline qp_qiime2/__init__.py:150]: This is also new, in the past, we will still run the code within the if, even if `add_method ` was already set as `False` so now skipping.

File: qp_qiime2/__init__.py [hunk 10]
```diff
@@ -76,234 +86,191 @@
...
-mid = m.id
+        # the diversity methods can have a choice param with no values
+        # so we need to fix so users can actually select things;
+        # however,we want to make sure that this is the only one,
+        # if not, raise an error
+        if data_type == 'choice' and default is None:
+            # these are valid optional parameters
+            valid_none_params = [
+                ('emperor', 'plot'),
+                ('gneiss', 'ilr_phylogenetic_ordination')]
```
Comment (antgonza) [inline qp_qiime2/__init__.py:236]: Finally, the other "important" change is that we add `('gneiss', 'ilr_phylogenetic_ordination')` to here as this is also a valid None value (optional)

**Review by antgonza — COMMENTED** [review]
Added some comments here to "facilitate" review; let me know if they help.

**Review by ElDeveloper — APPROVED** [review]
Thanks @antgonza

**Comment by ElDeveloper** [comment]
@antgonza I just restarted the jobs to see if the build passes now.

**Comment by antgonza** [comment]
Thank you; looks like tests passed.

**Comment by ElDeveloper** [comment]
Thanks @antgonza

**Pull request merged**